src/main/scala/org/broadinstitute/dsde/workbench/sam/directory/LdapDirectoryDAO.scala

```diff
@@ -41,59 +48,72 @@ class LdapDirectoryDAO(protected val ldapConnectionPool: LDAPConnectionPool, pro
     } *> IO.pure(group)
   }
 
-  override def loadGroup(groupName: WorkbenchGroupName): IO[Option[BasicWorkbenchGroup]] = cs.evalOn(ecForLdapBlockingIO) {
-    IO(Option(ldapConnectionPool.getEntry(groupDn(groupName))) map(unmarshalGroup))
+  override def loadGroup(groupName: WorkbenchGroupName): IO[Option[BasicWorkbenchGroup]] = {
+    for {
+      searchResult <- cs.evalOn(ecForLdapBlockingIO)(IO(ldapConnectionPool.getEntry(groupDn(groupName))))
+      res <- Option(searchResult).traverse{
+        entry =>
+          unmarshalGroup(entry).fold(err => IO.raiseError(new WorkbenchException(err)), r => IO.pure(r))
+      }
+    } yield res
```

Comment:
1. reason why I made the left type `String` here is because I want to combine these error messages into one easily (if it were Exception, I'd have to write a Semigroup instance for exception to combine them).

2. I think by promoting throwing exceptions, we have to code in a more defensive way and more mentally exhausting to me (have to think about whether a statement will throw all the time and make the decision whether we care to interpret it before propagating it up)

Reply:
I disagree on 2. I find it freeing. We thought about our global exception handler. Now I don't think about exceptions at all because I know they are handled. I only think about exceptions when I know I need to do some special handling which is rare.

Including exceptions in the return type of functions is the equivalent to checked exceptions in java. It forces us to deal with them at every level of the application. So I would argue it is more mental burden.